Refactor scil_compute_sf_from_sh.py

ppoulin91 opened the pull request:
Using Dipy's `sh_to_sf` caused memory problems because everything is computed in float64. I refactored the code to work like `convert_sh_basis`, so it can run in parallel, and I use float32 to make it fit in memory, especially when working with 1x1x1 volumes with > 200 SF directions.

pep8speaks commented:
Hello @ppoulin91, Thank you for submitting the Pull Request !

* In the file [`scilpy/reconst/multi_processes.py`](https://github.com/scilus/scilpy/blob/8bcf7e63152978a3e8287c58d9d6a8df52e9254b/scilpy/reconst/multi_processes.py):

> [Line 524:52](https://github.com/scilus/scilpy/blob/8bcf7e63152978a3e8287c58d9d6a8df52e9254b/scilpy/reconst/multi_processes.py#L524): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 524:80](https://github.com/scilus/scilpy/blob/8bcf7e63152978a3e8287c58d9d6a8df52e9254b/scilpy/reconst/multi_processes.py#L524): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)

CHrlS98 commented on scilpy/reconst/multi_processes.py:
```diff
@@ -472,3 +472,83 @@ def convert_sh_basis(shm_coeff, sphere, mask=None,
+    return chunk_id, sf
+
+
+def convert_sh_to_sf(shm_coeff, sphere, mask=None,
```
Since DIPY 's `sh_to_sf_matrix` supports it, maybe we could have an argument for `full_basis`?

CHrlS98 reviewed (commented):
Alright!
A small comment, but otherwise I like it.

Also I'm wondering if the type could also be an argument to `convert_sh_to_sf` if ever we'd want to use `float64`?

frheault commented on scripts/scil_compute_sf_from_sh.py:
```diff
@@ -43,14 +44,15 @@ def _build_arg_parser():
                    help="Generate a DWI-like output, including a `.bval` file "
                         "and b0 images in the sf file.")
-    p.add_argument('--bval', required=True,
+    p.add_argument('--bval',
```
I really don't undestand why bvecs is automatically saved (with a default filename we can't change), but bval is optional and we can choose the name.

frheault commented on scripts/scil_compute_sf_from_sh.py:
```diff
@@ -43,14 +44,15 @@ def _build_arg_parser():
     p.add_argument("--extract_as_dwi", action="store_true",
```
Also, I dont understand why extract_as_dwi is a parameters (it confused me at first).
Why not remove this parameters and simply say in --in_bval that the resulting SF are actually interpretable as DWI.
(It would be important to say that this is a 'related' script to scil_compute_sh_from_signal.py)

Also after that you can keep a warning for --b0 if --in_bval isn't there
Then --out_bvecs and --out_bvals could be both optional if the person want it or not.

frheault reviewed (commented):
I have a question, why only accept a sphere ?
It would be nice to have a mutually exclusive option to give either a sphere name or a bvecs so you can resample from 20 directions to 64 and not just 128 or 362 or 724.

(This is related to my suggestion that --out_bvecs and --out_bvals should all be paramaters)

ppoulin91 commented:
@CHrlS98 : I added a dtype option (limited to float32/float64), and a --full_basis option to support asym. SH. Let me know if this fits your needs and if anything needs changing!

@frheault : You're right, it was getting confusing. I removed the --extract_as_dwi option, and added --in_* and --out_* options which are all handled separately, and it's the user's responsibility to ask for the right outputs as needed. I also added the --in_bvec option (mutually exclusive with --sphere) to use arbitrary SF directions. Let me know what you think of the new version.

CHrlS98 reviewed (approved):
Alright, 
I just tried it on full SH basis and it works nicely.
Thanks!

ppoulin91 commented:
@frheault Some of your comments apply to functions out of the scope of the PR (I think the code was auto-formatted and I didn't notice) which are not in the script directly, so no access to `parser` and `args`. I'll fix it for my function+script, the rest should be done in another PR if we want to refactor stuff (and the `validate_nbr_processes()` function needs a touch up too, it has an unused argument and the docstring is not standard)

scil-admin commented:
Build passed ! Good Job :beers: !
